pkg/clustermesh/remote_cluster.go

```diff
@@ -134,26 +134,30 @@ func (rc *remoteCluster) releaseOldConnection() {
 	rc.backend = nil
 	rc.mutex.Unlock()
 
-	// Release resources asynchroneously in the background. Many of these
-	// operations may time out if the connection was closed due to an error
-	// condition.
-	go func() {
-		if ipCacheWatcher != nil {
-			ipCacheWatcher.Close()
-		}
-		if remoteNodes != nil {
-			remoteNodes.Close(context.TODO())
-		}
-		if remoteIdentityCache != nil {
-			remoteIdentityCache.Close()
-		}
-		if remoteServices != nil {
-			remoteServices.Close(context.TODO())
-		}
-		if backend != nil {
-			backend.Close()
-		}
-	}()
+	// The following caches must be closed synchronously to ensure that
+	// delete events are processed before the new connection is established
+	// which will trigger events of the new state again
+	if remoteNodes != nil {
+		remoteNodes.Close(context.TODO())
+	}
+	if remoteServices != nil {
+		remoteServices.Close(context.TODO())
+	}
+	if remoteIdentityCache != nil {
+		remoteIdentityCache.Close()
+	}
+	if ipCacheWatcher != nil {
+		ipCacheWatcher.Close()
+		ipCacheWatcher.DeleteEntries()
+	}
+	rc.mesh.globalServices.onClusterDelete(rc.name)
+
+	// Release the backend asynchronously in the background. Revoking
+	// leases and sessions may block. Success of this operation is not a
+	// pre-requisite for a successful reconnect.
+	if backend != nil {
+		go backend.Close()
```

Comment:
What happens if the `Close` blocks for a long time? Is there a scenario where there are many connections reestablished, hang / lose connectivity such that we need to start a new one, then are sent off to asynchronously close? Could these build up, and if so, should we track them or have a timeout to forcefully remove ourselves?

Not too familiar with the code here, but it would be worth explaining in a comment at least why we don't care about this goroutine.